### PR TITLE
[BugFix] Fix backend number assert

### PR DESCRIPTION
**Pull request by liuyehcf**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes 
1. https://github.com/StarRocks/StarRocksTest/issues/464
2. https://github.com/StarRocks/StarRocksTest/issues/471

## Problem Summary(Required) ：

Considering the following execution sequences

1. process fragment 0, which contains 1 instances, it's `backend_num=0`
    1. add `backend_num=0`  to `backendExecStates` of which size is up to 1
3. process fragment 1, which contains 3 instances, `backend_num` range from `1` to `3`. And the these 3 instances are divided into two parts,  first part contains `backend_num=2` and `backend_num=3`, second part contains `backend_num=1`
    1. we first process the first part, add `backend_num=2` and `backend_num=3` to `backendExecStates` of which size is up to 3
    3. execute `instance(backend_num=2)` and `instance(backend_num=3)` asynchronously
    4. `instance(backend_num=3)` is finished very soon, but the exec state is discarded, because the assertion `params.backend_num < backendExecStates.size()` is not satisfied
    5. then we process the second part, add `backend_num=1` to `backendExecStates` of which size is up to 4
    6. other steps...

Here are the logs corresponding to the above explanation.
```
2022-06-02 12:11:44,894 WARN (starrocks-mysql-nio-pool-0|99) [Coordinator.exec():401] exec begin, this=com.starrocks.qe.Coordinator@25b7e1c5
2022-06-02 12:11:44,895 WARN (starrocks-mysql-nio-pool-0|99) [Coordinator.exec():558] put backendExecStates backend_num=0, this=com.starrocks.qe.Coordinator@25b7e1c5
2022-06-02 12:11:44,897 WARN (starrocks-mysql-nio-pool-0|99) [Coordinator.exec():558] put backendExecStates backend_num=2, this=com.starrocks.qe.Coordinator@25b7e1c5
2022-06-02 12:11:44,897 WARN (starrocks-mysql-nio-pool-0|99) [Coordinator.exec():558] put backendExecStates backend_num=3, this=com.starrocks.qe.Coordinator@25b7e1c5
2022-06-02 12:11:44,907 WARN (starrocks-mysql-nio-pool-0|99) [Coordinator.exec():558] put backendExecStates backend_num=1, this=com.starrocks.qe.Coordinator@25b7e1c5
2022-06-02 12:11:44,909 WARN (starrocks-mysql-nio-pool-0|99) [Coordinator.exec():617] exec end, this=com.starrocks.qe.Coordinator@25b7e1c5
```

So, the solution is straightforward, we only need to check whether the reported `backend_num` exists in the `backendExecStates` instead of assuming the size relation.

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java [hunk 1]
```diff
@@ -1590,9 +1591,9 @@ private void computeScanRangeAssignment() throws Exception {
     }
 
     public void updateFragmentExecStatus(TReportExecStatusParams params) {
-        if (params.backend_num >= backendExecStates.size()) {
-            LOG.warn("unknown backend number: {}, expected less than: {}",
-                    params.backend_num, backendExecStates.size());
+        if (!backendExecStates.containsKey(params.backend_num)) {
```
Comment (mofeiatwork) [inline fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java:1594]: Why not merge it into the following `backendExecStates.get()` ?

**Comment by liuyehcf** [comment]
run starrocks_fe_unittest

**Review by satanson — APPROVED** [review]

**Comment by satanson** [comment]
"Backend number may be not continuous, so we cannot assert that be number is smaller than the size"

backend numbers must be continuous,  may be some upper fragments finished early because of short-circuit.

**Comment by liuyehcf** [comment]
> "Backend number may be not continuous, so we cannot assert that be number is smaller than the size"
> 
> backend numbers must be continuous, may be some upper fragments finished early because of short-circuit.

yes, you're right

**Review by kangkaisen — APPROVED** [review]

**Comment by liuyehcf** [comment]
@Mergifyio backport branch-2.2

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 4 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|4|00.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 0 / 4 (00.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|0|4|00.00%|

**Pull request merged**